includes/panopto_derivatives.inc

```diff
@@ -5,37 +5,356 @@
+  if ($response) {
+    $session = $response;
+  }
+  return $session;
+}
+
+/*
+ * @param $endpoint
+ * @param $headers
+ * @param $content
+ * @param $type
+ * @param $expect
+ */
+function islandora_video_streaming_panopto_request_curl($endpoint, $headers, $content = '', $type = 'GET', $expect = array()) {
+  $cURL = curl_init();
+  curl_setopt($cURL, CURLOPT_URL, $endpoint);
+  curl_setopt($cURL, CURLOPT_RETURNTRANSFER, true);
+  curl_setopt($cURL, CURLOPT_HEADER, 1);
+  curl_setopt($cURL, CURLINFO_HEADER_OUT, true);
+  switch (strtoupper($type)) {
+    case 'POST':
+    curl_setopt($cURL, CURLOPT_POST, true);
+    break;
+    case 'PUT':
+    curl_setopt($cURL, CURLOPT_CUSTOMREQUEST, 'PUT');
+    break;
+  }
+  if ($content) {
+    curl_setopt($cURL, CURLOPT_POSTFIELDS, $content);
+  }
+  curl_setopt($cURL, CURLOPT_HTTPHEADER, $headers);
+  $response = curl_exec($cURL);
+  $header_size = curl_getinfo($cURL, CURLINFO_HEADER_SIZE);
+  $response_code = curl_getinfo($cURL, CURLINFO_RESPONSE_CODE);
+  $headerSent = curl_getinfo($cURL, CURLINFO_HEADER_OUT );
+  curl_close($cURL);
+  if ($response === false) {
+    $messages = array(
+      'message' => t("Unable to receive response from curl_exec() for object @pid"),
+      'message_sub' => array('@pid' => $object->id),
+      'type' => 'watchdog',
+      'severity' => WATCHDOG_ERROR,
+    );
+    $failureMessage['messages'][] = $messages;
+  } else {
+    $headers = substr($response, 0, $header_size);
+    $body = substr($response, $header_size);
+    if (empty($expect) || (isset($expect[$response_code]) && empty($expect[$response_code]))) {
+      return $body;
+    } else if (isset($expect[$response_code])) {
+      $expectation = $expect[$response_code];
+      $json = json_decode($body, true);
+      if (is_array($json)) {
+        if (isset($json[$expectation])) {
+          return $json;
+        } else {
+          $messages = array(
+            'message' => t("For object @pid : JSON key from $endpoint ($expectation) missing . $response"),
+            'message_sub' => array('@pid' => $object->id),
+            'type' => 'watchdog',
+            'severity' => WATCHDOG_ERROR,
+          );
+          $failureMessage['messages'][] = $messages;
+        }
+      } else {
+        $messages = array(
+          'message' => t("For object @pid : Expected JSON key from $endpoint but could not decode. $response"),
+          'message_sub' => array('@pid' => $object->id),
+          'type' => 'watchdog',
+          'severity' => WATCHDOG_ERROR,
+        );
+        $failureMessage['messages'][] = $messages;
+      }
+    } else {
+      $messages = array(
+        'message' => t("For object @pid : Expected one of " .implode(',', array_keys($expect))."; got a $response_code"),
+        'message_sub' => array('@pid' => $object->id), 
+        'type' => 'watchdog',
+        'severity' => WATCHDOG_ERROR,
+      );
+      $failureMessage['messages'][] = $messages;
+    }
+  }
+  return;
```

Comment:
You've constructed a number of failure messages for `islandora_video_streaming_panopto_request_curl`, but none of them are used.  They should be logged or returned.

Reply:
Done. `failuremessages` are now returned.

Reply:
Returning the `$failureMessage` could be hazardous to this function.  I haven't looked at your implementation yet, but because this was originally written to return the successful result or false on error, the implementation of returning both successful content in tandem with failures will be complex.

Reply:
Would it be better to log the messages and have the functions that call `islandora_video_streaming_panopto_request_curl` to check whether the returned result was successful or not?

Reply:
Probably.  The downside to this is that you then can't pass specific messages back out to the end user.  The upside is that for cURL messages, you probably don't want specific messages to go to the user - they will be too technical.  Another user-facing message can be constructed based on a false result returned from the call.